File: src/py/ftp_init.py
import argparse
import sqlite3
import configparser
from simple_chalk import chalk, green, yellow, blue, red
from ftplib import FTP
from datetime import datetime

#############################################
## FUNCTIONS
#############################################

## INITIALISE DB Connection
def init_db(conn):

  c = conn.cursor()

  # Create ftp files table
  c.execute(
	'''
	CREATE TABLE IF NOT EXISTS pubmed_ftp (
	  name TEXT,
	  host_dir TEXT,
	  size INT,
		size_dl INT,
	  retrieved TEXT,
	  validated TEXT,
	  loaded TEXT,
	  pubtype TEXT,
		sts TEXT,
		err TEXT,
	  xmlBlob BLOB,
	  PRIMARY KEY (host_dir, name)
	);
	''')

## CHECK FOR FILES
def chk_pubmed(config, host_dir, pubtype, conn):
	print(blue('CHECKING'), yellow(pubtype), blue('FILES'))
	
	host = config.get('pubmed', 'ftp_host')
	user = config.get('pubmed', 'ftp_user')
	password = config.get('pubmed', 'ftp_password')
	
	c = conn.cursor()
	ftp = FTP(host)
	ftp.set_pasv(False)
	
	ftp.login(user=user, passwd=password)
	ftp.cwd(host_dir)
	
	size = 0
	for name, facts in ftp.mlsd():
		# print("name "+name+" size "+facts["size"])
		file_size = int(facts["size"])
		size += int(facts["size"])
		sql = '''
		  INSERT OR IGNORE INTO pubmed_ftp
			(name, host_dir, pubtype, size) 
		  VALUES("{name}", "{host_dir}", "{pubtype}", {size});
		'''.format(name=name,host_dir=host_dir,pubtype=pubtype,size=file_size) 
		c.execute(sql)      
		conn.commit()   
	
	ftp.quit()
	c.close()
	print(blue('Total files size'), yellow(str(size/1073741824)), blue('GB'))
	
	return

#############################################
## MAINLINE
#############################################

def main():
	config = configparser.ConfigParser()
	config.read("pubmed-db.cnf")
	
	parser = argparse.ArgumentParser()
	parser.add_argument("pubtype",help = "Nominate pubtype dowloads to be checked: BASE | UPDATE (case insensitive)")
	args = parser.parse_args()
	
	pubtype = str(args.pubtype).upper()

	if pubtype != 'BASE' and pubtype != 'UPDATE':
		print('invalid pubtype value', pubtype)
		return 
		
	if pubtype == 'BASE':
		host_dir = config.get('pubmed', 'base_folder')
		
	if pubtype == 'UPDATE':
		host_dir = config.get('pubmed', 'updates_folder')
	
	# Setup DB 
	controlDB = config.get('data', 'control_db')
	conn = sqlite3.connect(controlDB)
	init_db(conn)
	
	# check to see what files need to be downloaded
	chk_pubmed(config, host_dir, pubtype, conn)
	
	conn.close()
	return 

if __name__ == "__main__":
	main()

File: src/js/ftp_init.js
require('dotenv').config()
const yellow = require('chalk').yellow
const blue = require('chalk').blue
const ftp = require("basic-ftp")
const Database = require('better-sqlite3')

const BASE = 'BASE'
const UPDATE = 'UPDATE' 

main()

async function main () {
	
	var args = process.argv.slice(2)

	if (args.length == 0) {
		console.log("pubtype value not supplied. requires: BASE or UPDATE")
	} else {
		pubtype = args[0].toUpperCase()
		if (pubtype != BASE && pubtype != UPDATE) {
			console.log("pubtype value must be BASE or UPDATE")
		} else {
			db_init()
			ftp_init(pubtype)
		}
	}
}

async function ftp_init (pubtype) {

	const download_dir = (pubtype === UPDATE) ? process.env.pubmed_db_ftp_updates : process.env.pubmed_db_ftp_base
	const client = new ftp.Client()
	client.ftp.verbose = false
	try {
		await client.access({
			host: process.env.pubmed_db_ftp_host,
			user: process.env.pubmed_db_ftp_user,
			password: process.env.pubmed_db_ftp_password,
			secure: false
		})
		
		const db = new Database(process.env.pubmed_db_test + 'pubmed.ftp.sqlite', { /* verbose: console.log */ })
		let size = 0
		await client.cd(download_dir)
		const contents = await client.list()
		for (let item of contents) {
			size = size + item.size
			if (item.type === 1) update_control(db, pubtype, download_dir, item.name, item.size)
		}
		db.close()
		console.log(blue(`Total ${yellow(pubtype)} files size`), yellow((size/1073741824).toFixed(2)), blue('GB'))
	}
	catch(err) {
		console.log(err)
	}
	client.close()
}

function update_control  (db, pubtype, download_dir, file_name, size) {
	let sql = `	
		INSERT OR IGNORE INTO pubmed_ftp
			(name, host_dir, pubtype, size) 
  		VALUES('${file_name}', '${download_dir}', '${pubtype}', ${size});
	`
	const res = db.prepare(sql).run()	
}

async function db_init () {
	
	const db = new Database(process.env.pubmed_db_test + 'pubmed.ftp.sqlite', { /* verbose: console.log */ })
	let sql = `	
		CREATE TABLE IF NOT EXISTS pubmed_ftp (
		  name TEXT,
		  host_dir TEXT,
		  size INT,
			size_dl INT,
		  retrieved TEXT,
		  validated TEXT,
		  loaded TEXT,
		  pubtype TEXT,
			sts TEXT,
			err TEXT,
		  xmlBlob BLOB,
		  PRIMARY KEY (host_dir, name)
		);
	`
	const res = db.prepare(sql).run()	
	db.close()
}